quesma/v2/core/quesma_builder.go

```diff
@@ -59,52 +60,104 @@ func (quesma *Quesma) Stop(ctx context.Context) {
+			})
+			quesma.dependencies.Logger().Info().Msg(conn.InstanceName() +
+				":" + conn.GetEndpoint() +
+				",index:" + strconv.FormatInt(int64(connIndex), 10) +
+				",pipeline:" + strconv.FormatInt(int64(pipelineIndex), 10))
+		}
+	}
 
-	endpoints := make(map[string]struct{})
-	handlers := make(map[string]HandlersPipe)
+	// Second pass is about connecting routers with processors
+	// and merge them if they are the same properties
+	handlersPerEndpoint := make(map[string]map[string]HandlersPipe)
 	for _, pipeline := range quesma.pipelines {
 		for _, conn := range pipeline.GetFrontendConnectors() {
 			if httpConn, ok := conn.(HTTPFrontendConnector); ok {
-				endpoints[conn.GetEndpoint()] = struct{}{}
 				router := httpConn.GetRouter()
 				for path, handlerWrapper := range router.GetHandlers() {
 					handlerWrapper.Processors = append(handlerWrapper.Processors, pipeline.GetProcessors()...)
-					handlers[path] = handlerWrapper
+					sharedWrappers := make(map[string]HandlersPipe)
+					sharedWrappers[path] = handlerWrapper
+					if _, ok := handlersPerEndpoint[conn.GetEndpoint()]; ok {
+						for path, handlerWrapper := range handlersPerEndpoint[conn.GetEndpoint()] {
+							sharedWrappers[path] = handlerWrapper
+						}
+					}
+					handlersPerEndpoint[conn.GetEndpoint()] = sharedWrappers
 				}
 			}
 		}
 	}
-	if len(endpoints) == 1 {
-		for _, pipeline := range quesma.pipelines {
-			for _, conn := range pipeline.GetFrontendConnectors() {
-				if httpConn, ok := conn.(HTTPFrontendConnector); ok {
-					router := httpConn.GetRouter().Clone().(Router)
-					router.SetHandlers(handlers)
-					httpConn.AddRouter(router)
-				}
+
+	quesma.dependencies.Logger().Debug().Msg(dumpEndpoints(endpoints))
+
+	// This pass sets the routers with the handlers
+	for _, pipeline := range quesma.pipelines {
+		for _, conn := range pipeline.GetFrontendConnectors() {
+			if httpConn, ok := conn.(HTTPFrontendConnector); ok {
+				router := httpConn.GetRouter().Clone().(Router)
+				router.SetHandlers(handlersPerEndpoint[conn.GetEndpoint()])
+				httpConn.AddRouter(router)
 			}
 		}
-		// TODO this fixes the problem of sharing the same frontend connector
-		// in the case of having only one endpoint
-		// however it's not fully generic yet as only subset of connectors might be shared
-		if len(quesma.pipelines) > 1 {
-			if len(quesma.pipelines[0].GetFrontendConnectors()) == 0 {
-				return nil, fmt.Errorf("no frontend connectors provided")
-			}
-			sharedFc := quesma.pipelines[0].GetFrontendConnectors()[0]
-			for index := 1; index < len(quesma.pipelines); index++ {
-				for indexFc := range quesma.pipelines[index].GetFrontendConnectors() {
-					quesma.pipelines[index].GetFrontendConnectors()[indexFc] = sharedFc
+	}
+
+	// This pass is about sharing frontend connectors
+	for endpoint, endpointInfo := range endpoints {
+		sharedFc := quesma.pipelines[endpointInfo[0].pipelineIndex].GetFrontendConnectors()[endpointInfo[0].connIndex]
```

Comment:
it may be that `endpointInfo[0]` is a safe call here, but perhaps we should check that anyways?

Reply:
Thx. I will add sanity check